### DISCUSSION TITLE
jboss-logging i18n not working with gradle

### DISCUSSION THREAD

**Discussion by alexpoletaev** [quarkusio/quarkus#38436] [Q&A] [answered]
### Describe the bug

I want to use localized messages for the exceptions and logging. Since jboss-logging is already included in Quarkus it's perfectly fits my requirements. I used [this](https://jboss-logging.github.io/jboss-logging-tools/) documentation. I created an empty Quarkus project from the [starter](https://code.quarkus.io/) and additionaly added this dependency in gradle.build file:
`compileOnly 'org.jboss.logging:jboss-logging-processor:2.2.1.Final'`
But it's not working, interface implementations are not compiled at all. And i'm getting exceptions.
But it works perfectly with maven, I just added this dependency in pom.xml and that's all:
```
<dependency>
    <groupId>org.jboss.logging</groupId>
    <artifactId>jboss-logging-processor</artifactId>
    <version>2.2.1.Final</version>
    <scope>provided</scope>
    <optional>true</optional>
</dependency>
```

### Expected behavior

Localized message is successfully printed to the concole

### Actual behavior

I'm getting exceptions:
```
Invalid logger interface com.example.i18n.Logging (implementation not found in QuarkusClassLoader:Quarkus Runtime ClassLoader: DEV restart no:0@47c64cfe)
```
```
Invalid bundle interface com.example.i18n.Exceptions (implementation not found)
```

### How to Reproduce?

1) Start attached project (`quarkus-jb-gradle` is not working, while `quarkus-jb-maven` works fine)
[quarkus-jb-gradle.zip](https://github.com/quarkusio/quarkus/files/14066797/quarkus-jb-gradle.zip)
[quarkus-jb-maven.zip](https://github.com/quarkusio/quarkus/files/14066798/quarkus-jb-maven.zip)

2) execute `curl --request GET --url 'http://localhost:8080/hello'` or `curl --request GET --url 'http://localhost:8080/hello?param=throw'`

[a comment by an automated account is not shown]

**Comment by gastaldi** [comment]
Take a look at how it is done here (Maven): https://github.com/quarkiverse/quarkus-ironjacamar/blob/main/runtime/pom.xml#L75

**Comment by alexpoletaev** [comment]
> Take a look at how it is done here (Maven): https://github.com/quarkiverse/quarkus-ironjacamar/blob/main/runtime/pom.xml#L75

Thank you for the reply, do you have a working example with **Gradle**? Because, as I mentioned early, I don't have problems with Maven, it works fine for me.

**Comment by gastaldi** [comment, marked as answer]
To configure the `jboss-logging-processor` annotation processor in Gradle, you need to add it to the `annotationProcessor` configuration in your `dependencies` block. This will ensure that the annotation processor is run during the compile phase of your build.

Here is how you can do it:

```groovy
dependencies {
    // other dependencies...
    annotationProcessor 'org.jboss.logging:jboss-logging:3.5.3.Final'
    annotationProcessor 'org.jboss.logging:jboss-logging-processor:2.2.1.Final'
    annotationProcessor 'org.jboss.logging:jboss-logging-annotations:2.2.1.Final'
}
```

This will ensure that the annotation processor is included in the classpath during the compile phase.

Here is my `build.gradle` file: 

```gradle
plugins {
    id 'java'
    id 'io.quarkus'
}

repositories {
    mavenCentral()
    mavenLocal()
}

dependencies {
    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
    implementation 'io.quarkus:quarkus-arc'
    implementation 'io.quarkus:quarkus-resteasy-reactive'
    testImplementation 'io.quarkus:quarkus-junit5'
    testImplementation 'io.rest-assured:rest-assured'
    annotationProcessor 'org.jboss.logging:jboss-logging:3.5.3.Final'
    annotationProcessor 'org.jboss.logging:jboss-logging-processor:2.2.1.Final'
    annotationProcessor 'org.jboss.logging:jboss-logging-annotations:2.2.1.Final'
}

group 'org.acme'
version '1.0.0-SNAPSHOT'

java {
    sourceCompatibility = JavaVersion.VERSION_17
    targetCompatibility = JavaVersion.VERSION_17
}

test {
    systemProperty "java.util.logging.manager", "org.jboss.logmanager.LogManager"
}
compileJava {
    options.encoding = 'UTF-8'
    options.compilerArgs << '-parameters'
}

compileTestJava {
    options.encoding = 'UTF-8'
}
```

**Comment by alexpoletaev** [comment]
> To configure the `jboss-logging-processor` annotation processor in Gradle, you need to add it to the `annotationProcessor` configuration in your `dependencies` block. This will ensure that the annotation processor is run during the compile phase of your build.
> 
> Here is how you can do it:
> 
> ```groovy
> dependencies {
>     // other dependencies...
>     annotationProcessor 'org.jboss.logging:jboss-logging:3.5.3.Final'
>     annotationProcessor 'org.jboss.logging:jboss-logging-processor:2.2.1.Final'
>     annotationProcessor 'org.jboss.logging:jboss-logging-annotations:2.2.1.Final'
> }
> ```
> 
> This will ensure that the annotation processor is included in the classpath during the compile phase.
> 
> Here is my `build.gradle` file:
> 
> ```groovy-gradle
> plugins {
>     id 'java'
>     id 'io.quarkus'
> }
> 
> repositories {
>     mavenCentral()
>     mavenLocal()
> }
> 
> dependencies {
>     implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
>     implementation 'io.quarkus:quarkus-arc'
>     implementation 'io.quarkus:quarkus-resteasy-reactive'
>     testImplementation 'io.quarkus:quarkus-junit5'
>     testImplementation 'io.rest-assured:rest-assured'
>     annotationProcessor 'org.jboss.logging:jboss-logging:3.5.3.Final'
>     annotationProcessor 'org.jboss.logging:jboss-logging-processor:2.2.1.Final'
>     annotationProcessor 'org.jboss.logging:jboss-logging-annotations:2.2.1.Final'
> }
> 
> group 'org.acme'
> version '1.0.0-SNAPSHOT'
> 
> java {
>     sourceCompatibility = JavaVersion.VERSION_17
>     targetCompatibility = JavaVersion.VERSION_17
> }
> 
> test {
>     systemProperty "java.util.logging.manager", "org.jboss.logmanager.LogManager"
> }
> compileJava {
>     options.encoding = 'UTF-8'
>     options.compilerArgs << '-parameters'
> }
> 
> compileTestJava {
>     options.encoding = 'UTF-8'
> }
> ```

Thank you very much for your reply! It works now, but, unfortunately, only with the default locale. I can't figure out where should I put translation files so the jboss logger catch them and compile locale specific implementations. I tried to place them in the package, where the interfaces are located (like it's suggested in the docs) and in the resources with the full package path (`src/main/resource/com/example/i18n`), like it worked with Maven, but with no luck. Could you please help with this?

**Comment by alexpoletaev** [comment]
> Thank you very much for your reply! It works now, but, unfortunately, only with the default locale. I can't figure out where should I put translation files so the jboss logger catch them and compile locale specific implementations. I tried to place them in the package, where the interfaces are located (like it's suggested in the docs) and in the resources with the full package path (`src/main/resource/com/example/i18n`), like it worked with Maven, but with no luck. Could you please help with this?

I think I've found a solution, but I'm not sure if it's a correct one. I created a `i18n` directory inside resources directory, put translations there and register a gradle task:
```
tasks.register('copyTranslations', Copy) {
    from layout.projectDirectory.dir("src/main/resources/i18n")
    include "*.properties"
    into layout.buildDirectory.dir("classes/java/main/com/example/i18n")
}
```
then made a `compileJava` task to depend on my task:
```
compileJava {
    dependsOn copyTranslations
    options.encoding = 'UTF-8'
    options.compilerArgs << '-parameters'
}
```
After that locale specific implementations have been successfully generated

[a comment by an automated account is not shown]

**Comment by alexpoletaev** [comment]
I'm attaching a working solution (just for the history): 
[quarkus-jb-gradle-fix.zip](https://github.com/quarkusio/quarkus/files/14077192/quarkus-jb-gradle-fix.zip)
